mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp

```diff
@@ -312,6 +297,35 @@ LogicalResult SYCLIDConstructorOp::verify() {
   return emitBadSignatureError(*this);
 }
 
+/// Abstract memref.memory_space_cast and supported memref.cast operations.
+///
+/// As only the first element pointed to by each pointer argument is used and
+/// memory space casts can be omitted, these casts can be abstracted.
+static bool removeOperandsMemoryCasts(MutableOperandRange args) {
+  bool mutated = false;
+  constexpr auto getCast = [](OpOperand &operand) -> CastOpInterface {
+    Operation *op = operand.get().getDefiningOp();
+    if (!op)
+      return nullptr;
+    if (isa<memref::MemorySpaceCastOp>(op))
+      return cast<CastOpInterface>(op);
```

Comment:
How do we avoid to transform the case where the address space cast is not between generic and some other space?

Reply:
(using a more concrete SYCL op would help here)